shell/platform/common/cpp/client_wrapper/include/flutter/event_channel.h

```diff
@@ -0,0 +1,182 @@
+ public:
+  // Creates an instance that sends and receives event handler on the channel
+  // named |name|, encoded with |codec| and dispatched via |messenger|.
+  EventChannel(BinaryMessenger* messenger,
+               const std::string& name,
+               const MethodCodec<T>* codec)
+      : messenger_(messenger), name_(name), codec_(codec) {}
+  ~EventChannel() = default;
+
+  // Prevent copying.
+  EventChannel(EventChannel const&) = delete;
+  EventChannel& operator=(EventChannel const&) = delete;
+
+  // Registers a stream handler on this channel.
+  // If no handler has been registered, any incoming stream setup requests will
+  // be handled silently by providing an empty stream.
+  void SetStreamHandler(const StreamHandler<T>& handler) const {
+    // TODO: The following is required when nullptr
+    // can be passed as an argument.
+    // if (!handler) { /* <= available for more than C++17 */
+    //   messenger_->SetMessageHandler(name_, nullptr);
+    //   return;
+    // }
+
+    const auto* codec = codec_;
+    const std::string channel_name = name_;
+    const auto* messenger = messenger_;
+    EventSinkImplementation* current_sink = current_sink_;
+    BinaryMessageHandler binary_handler = [handler, codec, channel_name,
+                                           current_sink, messenger](
+                                              const uint8_t* message,
+                                              const size_t message_size,
+                                              BinaryReply reply) mutable {
+      std::unique_ptr<MethodCall<T>> method_call =
+          codec->DecodeMethodCall(message, message_size);
+      if (!method_call) {
+        std::cerr << "Unable to construct method call from message on channel "
+                  << channel_name << std::endl;
+        return;
+      }
+
+      const std::string& method = method_call->method_name();
+      if (method.compare(kOnListenMethod) == 0) {
+        if (current_sink) {
+          std::cerr << "Failed to cancel existing stream: " << channel_name
+                    << std::endl;
+          handler.onCancel(method_call->arguments());
+          delete current_sink;
+          current_sink = nullptr;
+        }
+
+        current_sink =
+            new EventSinkImplementation(messenger, channel_name, codec);
+        handler.onListen(method_call->arguments(),
+                         static_cast<EventSink<T>*>(current_sink));
+
+        {
+          auto result = codec->EncodeSuccessEnvelope();
+          uint8_t* buffer = new uint8_t[result->size()];
+          std::copy(result->begin(), result->end(), buffer);
+          reply(buffer, result->size());
+          delete[] buffer;
+        }
+      } else if (method.compare(kOnCancelMethod) == 0) {
+        if (current_sink) {
+          handler.onCancel(method_call->arguments());
+
+          auto result = codec->EncodeSuccessEnvelope();
+          uint8_t* buffer = new uint8_t[result->size()];
```

Comment:
Note that you are leaking `buffer` here, demonstrating why using `new`/`delete` manually should be avoided whenever possible. (Separately from the fact that I don't think this buffer should exist in the first place.)

Reply:
These code will be removed in the first place.